caddyhttp/basicauth/basicauth.go

```diff
@@ -60,6 +61,10 @@ func (a BasicAuth) ServeHTTP(w http.ResponseWriter, r *http.Request) (int, error
 
 			// remove credentials from request to avoid leaking upstream
 			r.Header.Del("Authorization")
+
+			// let upstream middleware (e.g. fastcgi and cgi) know about authenticated
+			// user; this replaces the request with a wrapped instance
+			r = r.WithContext(context.WithValue(r.Context(), "remote_user", username))
```

Comment:
Instead of using a primitive string type as a context key (which is asking for collisions, though I admit, less likely in Caddy-specific packages), we should wrap this in `httpserver.CtxKey`. It doesn't exist yet but it will when #1430 is merged! I should have done that change separately, sorry. Also see #1331 which may be blocked on the same thing.